### PR TITLE
Improve how common whole number combinations are spoken

### PR DESCRIPTION
**Pull request by sclower**
## Updates how exponents are read aloud:

- x^1 = "x to the first power"
- x^2 = "x squared"
- x^3 = "x cubed"
- Other values are spoken as "to the y power', e.g. x^10 is "x to the 10th power." The suffix "st", "nd", etc is only applied to exponents whose absolute value is less than 1000. After some testing, larger values began to clash with screen reader number parsing (especially with 7-digit strings that VoiceOver read aloud as a phone number).
- Anything else is spoken as before: "x, superscript, expression, baseline."

## Updates Fraction Speech

- - If both numerator and denominator are numeric integers and denominator is between 2 and 9, speaks a simplified form e.g. "1 half" or "3 quarters."
- Fractions will otherwise be spoken as before, e.g. "Start Fraction, numerator over denominator, End Fraction."
- If a whole number is followed by a simple numeric fraction (e.g. 1\frac{1}{2}), we will insert "and" between the two so the mixed fraction is read more naturally, e.g. "1 and 1 half."

## Other updates
- Adds distinction for "minus" and "negative" symbols. e.g. 1-3 is read as "1 minus 3" whereas -1 is spoken as "negative 1."
- Adds distinction for "plus" and "positive" symbols. e.g. 2+2 is read as "2 + 2" whereas +1 is spoken as "positive 1."
- Ensures that mathspeak resulting from ARIA alerts is never shortened to avoid ambiguity. Specifically, if moving before or after a fraction or superscript, it is less confusing to be told about the item's technical form, while saying "after squared" or "after to the 10th power" is strange.
- [ ] tests pending behavioral approval

### CODE REVIEW

File: src/commands/math.js [hunk 1]
```diff
@@ -394,10 +395,23 @@ var MathBlock = P(MathElement, function(_, super_) {
         }
         var mathspeakText = cmd.mathspeak();
         var cmdText = cmd.ctrlSeq;
-        if (isNaN(cmdText) && cmdText !== '.') {
+        var isCmdNumeric = !isNaN(cmdText);
```
Comment (jwmerrill) [inline src/commands/math.js:398]: I think we want to be limiting this to only integers. For example, we don't want to speak `1.2 \frac{2}{3}` as "one point two and two thirds." The calculator treats this example by multiplying the numbers rather than adding them.

I'd also advocate for using regex to look for integers rather than `isNaN`, I think. I know that `isNaN` on a string will attempt to convert the string to a number first, but I think the details of exactly how that works are hard to guess and hard to remember.

File: src/commands/math.js [hunk 2]
```diff
@@ -394,10 +395,23 @@ var MathBlock = P(MathElement, function(_, super_) {
...
+
+        // Handle the case of a digit followed by a simplified fraction such as 1\frac{1}{2}.
+        // Such combinations should be spoken aloud as "1 and 1 half."
+        if (
+          wasPrevNumeric &&
+          cmdText.indexOf('frac') !== -1 &&
+        mathspeakText.indexOf('Fraction') === -1
```
Comment (jwmerrill) [inline src/commands/math.js:405]: This line should be indented two more spaces. Miss having prettier here ...

File: src/commands/math/commands.js [hunk 3]
```diff
@@ -342,7 +342,21 @@ LatexCmds['^'] = P(SupSub, function(_, super_) {
     + '</span>'
   ;
   _.textTemplate = [ '^' ];
-  _.mathspeakTemplate = [ 'Superscript,', ', Baseline'];
+  _.mathspeak = function() {
+    // Simplify basic exponent speech for common whole numbers.
+    var innerMathspeak = this.ends[L] && this.ends[L].mathspeak();
+    if (innerMathspeak === undefined) {
+      return '';
```
Comment (jwmerrill) [inline src/commands/math/commands.js:349]: We used to speak `x^{}+1` as "x superscript baseline plus 1" but I think this will now make us voice this as "x+1". The calculator treats this empty superscript as an error, so I think it probably makes sense for the mathspeak to somehow represent the difference between `x^{}+1` and `x+1`.

File: src/commands/math/commands.js [hunk 4]
```diff
@@ -342,7 +342,21 @@ LatexCmds['^'] = P(SupSub, function(_, super_) {
...
+    } else if (innerMathspeak === '1') {
+      return 'to the first power';
+    } else if (innerMathspeak === '2') {
+      return 'squared';
+    } else if (innerMathspeak === '3') {
+      return 'cubed';
+    }
```
Comment (jwmerrill) [inline src/commands/math/commands.js:356]: Should we also add 4-9 here?

Comment (sclower) [reply src/commands/math/commands.js:356]: After talking with Dave, we decided not to for now.

File: src/commands/math.js (hunk 1 above)
Comment (sclower) [reply src/commands/math.js]: I'm not quite sold on limiting the mixed fraction logic to entirely whole numbers, unless we find a way to also do away with the shortened speech for the fraction itself. e.g. 1.2 2 thirds sounds strange to me. 1.2 and 2 thirds might be unusual, but it's still understandable, as would be 1.2 start fraction, 2 over 3, end fraction.

Agreed re swapping out !isNaN with a regular expression.

File: src/commands/math/commands.js [hunk 5]
```diff
@@ -342,8 +346,55 @@ LatexCmds['^'] = P(SupSub, function(_, super_) {
     + '</span>'
   ;
   _.textTemplate = [ '^' ];
-  _.mathspeakTemplate = [ 'Superscript,', ', Baseline'];
+  _.mathspeak = function(opts) {
+    // Simplify basic exponent speech for common whole numbers.
+    var child = this.ends[L] || this.ends[R];
```
Comment (anandthakker) [inline src/commands/math/commands.js:351]: What's the difference between `this.ends[L]` and `this.ends[R]`?

File: src/commands/math/commands.js [hunk 6]
```diff
@@ -342,8 +346,55 @@ LatexCmds['^'] = P(SupSub, function(_, super_) {
...
+    if (child !== undefined) {
+      // Calculate this item's inner text to determine whether to shorten the returned speech.
+      // Do not calculate its inner mathspeak now until we know that the speech is to be truncated.
+      // Since the mathspeak computation is recursive, we want to call it only once in this function to avoid performance bottlenecks.
+      var innerText = typeof(child) === 'object'
+        ? child.text()
+        : ''+child;
```
Comment (anandthakker) [inline src/commands/math/commands.js:358]: I'm guessing that maybe the answer is "no", but I'm wondering if it's possible to answer the question of whether the exponent is a whole number by inspecting properties of the node rather than serializing it to text first?

File: test/unit/publicapi.test.js [hunk 7]
```diff
@@ -184,11 +184,11 @@ suite('Public API', function() {
       assertMathSpeakEqual(mq.mathspeak(), 'StartFraction "d" Over "d" "x" EndFraction StartRoot "x" EndRoot');
 
       mq.latex('1+2-3\\cdot\\frac{5}{6^7}=\\left(8+9\\right)');
-      assertMathSpeakEqual(mq.mathspeak(), '1 plus 2 minus 3 times StartFraction 5 Over 6 Superscript 7 Baseline EndFraction equals left parenthesis 8 plus 9 right parenthesis');
+      assertMathSpeakEqual(mq.mathspeak(), '1 plus 2 minus 3 times StartFraction 5 Over 6 to the 7 th power EndFraction equals left parenthesis 8 plus 9 right parenthesis');
```
Comment (anandthakker) [inline test/unit/publicapi.test.js:187]: Just checking: is the space between "7" and "th" intended/correct?

File: test/unit/typing.test.js [hunk 8]
```diff
@@ -75,6 +75,147 @@ suite('typing with auto-replaces', function() {
...
+
+    test('plus and minus differentiation', function() {
+      // Distinguish between positive vs plus and negative vs. minus
+      mq.latex('-25-25');
+      assertMathspeak('negative 25 minus 25');
+      mq.latex('+25+25');
+      assertMathspeak('positive 25 plus 25');
+    });
+  });
+
```
Comment (anandthakker) [inline test/unit/typing.test.js:218]: Nice tests

File: src/commands/math/commands.js [hunk 9]
```diff
@@ -165,6 +165,10 @@ var Class = LatexCmds['class'] = P(MathCommand, function(_, super_) {
   };
 });
 
+// This test is used to determine whether an item may be treated as a whole number
+// for shortening the verbalized (mathspeak) forms of some fractions and superscripts.
+var intRgx = new RegExp(/^[\+\-]?[\d]+$/);
```
[the remark this answers is not in the document]
Comment (sclower) [reply src/commands/math/commands.js:170]: Force of habit. Since we aren't dynamically creating the expression with a variable, I'll simplify.

File: src/commands/math/commands.js (hunk 5 above)
Comment (sclower) [reply src/commands/math/commands.js:351]: Good catch. Left this extra check in while debugging.

A Mathquill mode has a possible parent, sibling nodes, and child nodes. In this repo the "left" and "right" nodes are referenced with either node[L] or node[R] (L and R are constants of -1 and 1). The node.ends property contains child nodes. Some constructs (like fractions) have left and right children (numerator and denominator), but in the case of superscripts and subscripts the left and right children are the same, if they exist at all.

File: src/commands/math/commands.js (hunk 6 above)
Comment (sclower) [reply src/commands/math/commands.js:358]: Nope. Each digit is its own node so we have to walk the tree, even shallowly, no matter what.

File: test/unit/publicapi.test.js (hunk 7 above)
Comment (sclower) [reply test/unit/publicapi.test.js:187]: No. That's a bug.

File: src/commands/math/basicSymbols.js [hunk 10]
```diff
@@ -788,31 +788,34 @@ LatexCmds['¾'] = bind(LatexFragment, '\\frac34');
...
-        }
-      } else if (node.parent && node.parent.parent && node.parent.parent.isStyleBlock()) {
-        //if we are in a style block at the leftmost edge, determine unary/binary based on
-        //the style block
-        //this allows style blocks to be transparent for unary/binary purposes
-        return determineOpClassType(node.parent.parent);
-      } else {
-        return '';
+  _.init = VanillaSymbol.prototype.init;
+  _.isBinaryOperator = function(node) {
```
Comment (anandthakker) [inline src/commands/math/basicSymbols.js:793]: Nit: since `isBinaryOperator` does not appear to use `this`, I think it could just be a standalone function instead of a member method.

File: src/commands/math/commands.js (hunk 5 above)
Comment (anandthakker) [reply src/commands/math/commands.js]: > in the case of superscripts and subscripts the left and right children are the same, if they exist at all.

Is there a place in the code where I can see why/how that is the case?

Comment (sclower) [reply src/commands/math/commands.js]: The party starts in /src/tree.js.

**Comment by sclower** [comment]
retest this please

**Comment by sclower** [comment]
retest windows

File: src/commands/math/commands.js (hunk 5 above)
Comment (sclower) [reply src/commands/math/commands.js]: On closer inspection, superscripts are different. The finalizeTree methods in supsub and superscript should help.

File: src/commands/math/commands.js [hunk 11]
```diff
@@ -169,6 +169,28 @@ var Class = LatexCmds['class'] = P(MathCommand, function(_, super_) {
...
+  )
+    return block;
+  var children = block.children();
+  if (!children || !children.ends[L]) return block;
+  var chars = '';
+  for (var sibling = children.ends[L]; sibling[R] !== undefined; sibling = sibling[R]) {
+    if (sibling.ctrlSeq !== undefined) {
+      chars += sibling.ctrlSeq;
+    } else {
+      break;
```
Comment (anandthakker) [inline src/commands/math/commands.js:188]: When can `ctrlSeq` be undefined?  Wondering if this could result in a false positive, e.g. in a case where there are three children, the first being a digit, the next being a node with `ctrlSeq === undefined`, and the third being some non-digit node.  In that case, I think we'd just return the digit.

File: src/commands/math/commands.js [hunk 12]
```diff
@@ -165,6 +165,28 @@ var Class = LatexCmds['class'] = P(MathCommand, function(_, super_) {
...
+  if (
+    typeof(block) !== 'object' ||
+    typeof(block.children) !== 'function'
+  )
+    return block;
+  var children = block.children();
+  if (!children || !children.ends[L]) return block;
+  var chars = '';
+  for (var sibling = children.ends[L]; sibling[R] !== undefined; sibling = sibling[R]) {
+    chars += sibling.ctrlSeq;
```
Comment (jwmerrill) [inline src/commands/math/commands.js:185]: Is `ctrlSeq` always defined?

File: src/commands/math/commands.js [hunk 13]
```diff
@@ -461,14 +528,67 @@ LatexCmds.fraction = P(MathCommand, function(_, super_) {
...
+
+    var numText = getCtrlSeqsFromBlock(this.ends[L]);
+    var denText = getCtrlSeqsFromBlock(this.ends[R]);
+
+    // Shorten mathspeak value for whole number fractions whose denominator is less than 10.
+    if (
+      (!opts || !opts.ignoreShorthand) &&
+      intRgx.test(numText) && intRgx.test(denText)
+    ) {
+      var isSingular = Math.abs(numText) === 1;
```
Comment (jwmerrill) [inline src/commands/math/commands.js:552]: Mostly a style thing, but I think it's better not to rely in calling math functions like `Math.abs` on strings. Partly for readability, and partly because I'm never totally sure what will happen if the string isn't a number. This could be
```
numText === '1 || numText === '-1'
```
instead, right?

File: src/commands/math/commands.js [hunk 14]
```diff
@@ -165,6 +165,28 @@ var Class = LatexCmds['class'] = P(MathCommand, function(_, super_) {
   };
 });
 
+// This test is used to determine whether an item may be treated as a whole number
+// for shortening the verbalized (mathspeak) forms of some fractions and superscripts.
+var intRgx = /^[\+\-]?[\d]+$/;
+
+// Traverses the top level of the passed block's children and returns the concatenation of their ctrlSeq properties.
+// Used in shortened mathspeak computations as a block's .text() method can be potentially expensive.
```
Comment (jwmerrill) [inline src/commands/math/commands.js:173]: Nice idea on this strategy, and particularly avoiding making it recursive.

File: test/unit/typing.test.js [hunk 15]
```diff
@@ -75,6 +75,147 @@ suite('typing with auto-replaces', function() {
...
+      // A number followed by a shortened fraction should include the word "and", and other combinations should not.
+      mq.latex('3\\frac{3}{8}');
+      assertMathspeak('3 and 3 eighths');
+      mq.latex('3\\ \\frac{3}{8}');
+      assertMathspeak('3 and 3 eighths');
+      mq.latex('3\\frac{3}{x}');
+      assertMathspeak('3 StartFraction, 3 Over "x", EndFraction');
+      mq.latex('x\\frac{3}{8}');
+      assertMathspeak('"x" 3 eighths');
```
Comment (jwmerrill) [inline test/unit/typing.test.js:146]: Can we add a test that `1.2 \frac{1}{2}` is not spoke with "and" in it? I'm guessing it will be "1.2 1 half", which is a little weird, but still better than "1.2 and 1 half" because the latter implies addition rather than multiplication of these terms.

File: test/unit/typing.test.js [hunk 16]
```diff
@@ -75,6 +75,147 @@ suite('typing with auto-replaces', function() {
...
+      mq.latex('\\frac{x}{2}');
+      assertMathspeak('StartFraction, "x" Over 2, EndFraction');
+      mq.latex('\\frac{2}{x}');
+      assertMathspeak('StartFraction, 2 Over "x", EndFraction');
+
+      // Traditional fractions should be spoken if either numerator or denominator are not whole numbers
+      mq.latex('\\frac{1.2}{2}');
+      assertMathspeak('StartFraction, 1.2 Over 2, EndFraction');
+      mq.latex('\\frac{4}{2.3}');
+      assertMathspeak('StartFraction, 4 Over 2.3, EndFraction');
```
Comment (jwmerrill) [inline test/unit/typing.test.js:136]: Could we add a test that negative denominators are not shortened?

File: test/unit/typing.test.js [hunk 17]
```diff
@@ -75,6 +75,147 @@ suite('typing with auto-replaces', function() {
...
+      mq.latex('\\frac{31}{9}');
+      assertMathspeak('31 ninths');
+      mq.latex('\\frac{31}{10}');
+      assertMathspeak('StartFraction, 31 Over 10, EndFraction');
+
+      // Fractions with negative numerators should be shortened
+      mq.latex('\\frac{-1}{2}');
+      assertMathspeak('negative 1 half');
+      mq.latex('\\frac{-3}{2}');
+      assertMathspeak('negative 3 halves');
```
Comment (jwmerrill) [inline test/unit/typing.test.js:124]: Could we add a test for the case that the negative sign is outside the fraction?

File: test/unit/typing.test.js [hunk 18]
```diff
@@ -75,6 +75,147 @@ suite('typing with auto-replaces', function() {
...
+      mq.latex('x^{10000000000}');
+      assertMathspeak('"x" to the 10000000000 power');
+
+      // Ensure negative exponents are shortened
+      mq.latex('10^{-5}');
+      assertMathspeak('10 to the negative 5th power');
+      mq.latex('x^{-5}');
+      assertMathspeak('"x" to the negative 5th power');
+
+      // Superscripts that are not strictly integers should continue to be spoken in longer form
```
Comment (jwmerrill) [inline test/unit/typing.test.js:201]: Now wondering if we'll end up wanting to just always speak exponents in the "power" form. E.g it might be nice to have `e^{-x^2}` be "e to the negative x squared power".

I'm happy to land what we have and consider this as a later extension.

File: src/services/aria.js [hunk 19]
```diff
@@ -26,16 +26,30 @@ var Aria = P(function(_) {
   };
 
   _.queue = function(item, shouldDescribe) {
+    var output = '';
     if (item instanceof Node) {
+      // Some constructs include verbal shorthand (such as simple fractions and exponents).
+      // Since ARIA alerts relate to moving through interactive content, we don't want to use that shorthand if it exists
+      // since doing so may be ambiguous or confusing.
```
Comment (jwmerrill) [inline src/services/aria.js:33]: Do we have any test coverage of this distinction?

File: src/commands/math/commands.js (hunk 12 above)
Comment (sclower) [reply src/commands/math/commands.js:185]: I believe so. At least, I haven't seen any places in the code where we check for the existence of ctrlSeq. Wish this were strongly typed. I can add the check back in just to be safe.

File: src/commands/math/commands.js [hunk 20]
```diff
@@ -389,7 +389,8 @@ LatexCmds['^'] = P(SupSub, function(_, super_) {
         // More complex cases.
         var suffix = '';
         // Limit suffix addition to exponents < 1000.
-        if (Math.abs(innerText) < 1000) {
+        var innerNumber = parseInt(innerText, 10);
+        if (innerNumber !== NaN && Math.abs(innerNumber) < 1000) {
```
[the remark this answers is not in the document]
Comment (anandthakker) [reply src/commands/math/commands.js:393]: Alternatively, could also use a regexp here: I think `/^[+-]?\d{1,3}$/.test(innerText)` would do it?

File: src/commands/math/commands.js [hunk 21]
```diff
@@ -585,7 +586,24 @@ LatexCmds.fraction = P(MathCommand, function(_, super_) {
...
       if (newDenSpeech !== '') {
-        return this.ends[L].mathspeak() + ' ' + newDenSpeech;
+        var output = '';
+        // Handle the case of an integer followed by a simplified fraction such as 1\frac{1}{2}.
+        // Such combinations should be spoken aloud as "1 and 1 half."
+        // Start at the left sibling of the fraction and continue leftward until something other than a digit or whitespace is found.
+        var precededByInteger = false;
+        for (var sibling = this[L]; sibling[L] !== undefined; sibling = sibling[L]) {
+          if (sibling.ctrlSeq === '\\ ' || intRgx.test(sibling.ctrlSeq)) {
+            precededByInteger = true;
```
Comment (anandthakker) [inline src/commands/math/commands.js:596]: I believe this going to result in `'\\ \\frac{1}{2}'` being spoken as "and 1 half".  I think maybe what we want here is to `continue` in the whitespace case?

Comment (anandthakker) [reply src/commands/math/commands.js:596]: This also makes me think we should add some more test cases involving whitespace -- the problematic case above, maybe also e.g. `1\\ \\ \\frac{1}{2}` and `1.2\\ \\ \\frac{1}{2}`

**Comment by sclower** [comment]
Good observations. Code modified and tests added.

File: test/unit/aria.test.js [hunk 22]
```diff
@@ -38,6 +38,10 @@ suite('aria', function() {
     assertAriaEqual('over');
     mathField.typedText('2');
     assertAriaEqual('2');
+    mathField.keystroke('Tab');
+    assertAriaEqual('after StartFraction, 1 Over 2 , EndFraction');
+    mathField.keystroke('Backspace');
+    assertAriaEqual('end of denominator 2');
```
Comment (anandthakker) [inline test/unit/aria.test.js:44]: Maybe leave a brief comment here describing what this is testing?

**Review by anandthakker — APPROVED** [review]
